### ISSUE TITLE
Larastan doesn't detect custom model collection after `map()`

### ISSUE THREAD

**Issue by rudiedirkx** [nunomaduro/larastan#1927]
- Larastan Version: 2.9.5
- PHPStan Version: 1.10.67
- Laravel Version: 10.48.8

### Description

My `User` model overrides `newCollection()` to return custom `ModelCollection` that extends `Eloquent\Collection`. The user model has a relationship `created_by_user` to `User`. I want to fetch a few users, and map them to their creator users. The fetched users are a `ModelCollection` 👍 I expect the resulting collection to also be a `ModelCollection` because it's still a collection of models. Instead PHPStan thinks it's a Eloquent\Collection.

### Code

```php
$models = User::query()->get();
\PHPStan\dumpType($models); // 👍 App\Base\ModelCollection<int, App\Models\User>
$users = $models->map(fn(User $model) => $model->created_by_user);
\PHPStan\dumpType($users); // 👎 Illuminate\Database\Eloquent\Collection<int, App\Models\User>
```

If I make the closure more verbose:

```
$users = $models->map(function(User $model) : User {
    return $model->created_by_user;
});
```

the result is the same, because the `User` bit isn't the problem (Larastan knows the result is a collection of User), but the collection is.

### Demo

`MapController` in https://github.com/rudiedirkx/larastan-demo

It tests with `User` and `Post`. `User` has the custom collection, `Post` doesn't. The `Post` map result is fine, but the `User` map result is not.

    $ ./vendor/bin/phpstan analyze app/Http/Controllers/MapController.php

```
  16     Dumped type: Illuminate\Database\Eloquent\Collection<int, App\Models\Post>
  18     Dumped type: Illuminate\Database\Eloquent\Collection<int, App\Models\User>
  24     Dumped type: App\Base\ModelCollection<int, App\Models\User>
  28     Dumped type: Illuminate\Database\Eloquent\Collection<int, App\Models\User>
```

**Comment by rudiedirkx** [comment]
The weird union type with `Illuminate\Support\Collection<int, null>` only happens with the short closure I think:

    $users = $models->map(fn(Post $model) => $model->created_by_user);

The long closure results in a prettier `Illuminate\Database\Eloquent\Collection<int, App\Models\User>` but no `ModelCollection`.

**Comment by rudiedirkx** [comment]
I removed the weird union type bits, I don't care about those. Just use a long closure with return type. But the disappearing `ModelCollection` ... 😢

**Comment by calebdw** [comment]
Issue is that `static<...>` in the EloquentCollection stub is not supported: https://github.com/phpstan/phpstan/issues/5512

We would have to wait until this is implemented or create a dynamic return type extension for those methods

**Comment by canvural** [comment]
> Issue is that `static<...>` in the EloquentCollection stub is not supported: https://github.com/phpstan/phpstan/issues/5512
> 
> We would have to wait until this is implemented or create a dynamic return type extension for those methods

Yes,that's exactly what we already have. We are faking the generic static type.

**Comment by calebdw** [comment]
> Yes,that's exactly what we already have.

No, there's an extension for the SupportCollection, but the only extension for the EloquentCollection is for the `find` method

**Comment by canvural** [comment]
> > Yes,that's exactly what we already have.
> 
> No, there's an extension for the SupportCollection, but the only extension for the EloquentCollection is for the `find` method

The enumerable generic static extension is for this purpose.

**Comment by calebdw** [comment]
Yes, it just needs to be updated for the map* methods on the EloquentCollection

**Comment by canvural** [comment]
> Yes, it just needs to be updated for the map* methods on the EloquentCollection

No it's already doing it. It just has a bug.

**Comment by calebdw** [comment]
So only these static methods are supported:

https://github.com/larastan/larastan/blob/92376b5b7ed1f34cdfef92fc685dcd94348c8e65/src/ReturnTypes/EnumerableGenericStaticMethodDynamicStaticMethodReturnTypeExtension.php#L39-L47

And only the `find` method is currently supported for the EloquentCollection:

https://github.com/larastan/larastan/blob/92376b5b7ed1f34cdfef92fc685dcd94348c8e65/src/ReturnTypes/EnumerableGenericStaticMethodDynamicMethodReturnTypeExtension.php#L36-L40

**Comment by rudiedirkx** [comment]
Wooow I found an actual bug? I'm kinda proud of that. Larastan has been stellar, it's me with the bugs usually. Great work once again guys ❤️

**Comment by canvural** [comment]
Fixed with 1a05ac6dd0bbde4dc52b9d24c0a2a493f2c3784c
